event_listener.go

```diff
@@ -106,3 +112,18 @@ func (el *EventListener) Parse(l types.Log) *ContractEvent {
 		Removed:     l.Removed,
 	}
 }
+
+func (el *EventListener) channelCleanUp() {
+	// Unsubscribe should be called before this cleanUp stage, therefore geth
+	// would stop sending logs through the log channel (but it won't close it).
+	// The goal of this function is to drain the log channel, send events through
+	// event channel and close it (to notify receiver that there's no more data).
+	for len(el.logCh) > 0 {
+		log := <-el.logCh
+		if cEvent := el.Parse(log); cEvent != nil {
+			el.eventCh <- cEvent
+		}
+	}
+	close(el.eventCh)
```

Comment:
So we don't wait until the events are processed?

Reply:
Receiver can still get data from channel even after sender closing it. The reason that sender HAS to close channel is to let receiver know there is nomore data.

Reply:
`Receiver can still get data from channel even after sender closing it.`
I don't think so. @tailingchen how about your thoughts?